### PR TITLE
Fix/multilabel confusion matrix

### LINKED ISSUE #50: bug: multilabel confusion matrix
image

### PR DESCRIPTION
**Pull request by tiaguinho-code**
# What it do?
Give dataloader capability to decode target arrays from test and prediction set.
Should fix Issue #50 .

# How?
Added internal function to dataloader that usess the same label binarizer that it used to encode the different categories.
It does the inverse encoding then and returns that for an arbitrary target array.

In the current implementaiton the dataloader is passed to the evaluation script to do translation in there but we can of course also translate inside the trainer script and pass the decoded arrays and labels.

### CODE REVIEW

**Comment by tiaguinho-code** [comment]
Just noticed there is undefined behavior if the Model Chooses multiple elements for one row for example  THF and Et2O on the same molecule. Looks like the Binarizer somehow chooses one, but I guess this shouldn't be an issue as it will work out in the statistics.

**Comment by mlederbauer** [comment]
Thanks a lot for the contribution! Checking the code asap
I was wondering if you could add 1 or 2 example confusion matrices that are generated when running the code? (I guess this might change a bit wth our discussion on Discord)
Typically, you can “Copy Image” directly from VSCode and Cmd+V it here on Github.

**Comment by tiaguinho-code** [comment]
willco tomorrow 👁️

**Comment by tiaguinho-code** [comment]
willco tomorrow 👍🏼

**Comment by tiaguinho-code** [comment]
# Current state
The code has now been refactored quite a bit, especially the training script. It can handle all sorts of dataset mixtures. Roc etc only are generated for binary tasks and depending on the target dimensionality, different things are executed. 
# Todo
Fix the algorithm to make the subplots and add that cool color bar, but I feel like this depends on how exactly we want to plot it. The technically painful part is done, which is generate all the sub cms and bring along the right labels.
# Examples of generated confusion matrices
<details>
<summary>Images</summary>

image

image

image
</details>

**Review by tiaguinho-code — COMMENTED** [review]
.

File: nmrcraft/analysis/plotting.py [hunk 1]
```diff
@@ -84,29 +84,60 @@ def plot_predicted_vs_ground_truth_density(
...
+def plot_confusion_matrix(
+    cm, classes, title, path, full=True, columns_set=False
+):
     """
     Plots the confusion matrix.
     Parameters:
     - cm (array-like): Confusion matrix data.
     - classes (list): List of classes for the axis labels.
     - title (str): Title of the plot.
+    - full (bool): If true plots one big, else many smaller.
```
Comment (mlederbauer) [inline nmrcraft/analysis/plotting.py:96]: I like the description hehe

Comment (mlederbauer) [reply nmrcraft/analysis/plotting.py:96]: Alternatively for the final review round in the end, we could rename to “Plots single confusion matrix for all targets combined vs one CM per target” or something like that

File: nmrcraft/analysis/plotting.py [hunk 2]
```diff
@@ -84,29 +84,60 @@ def plot_predicted_vs_ground_truth_density(
...
+        tick_marks = np.arange(len(classes))
+        plt.xticks(tick_marks, classes, rotation=45)
+        plt.yticks(tick_marks, classes)
+        plt.tight_layout()
+        plt.ylabel("True label")
+        plt.xlabel("Predicted label")
+        plt.savefig(path)
+        plt.close()
+
+    elif not full:  # Plot many small cms of each target
```
Comment (mlederbauer) [inline nmrcraft/analysis/plotting.py:116]: could rewrite this to `else:` as full is either True or False

File: nmrcraft/analysis/plotting.py [hunk 3]
```diff
@@ -84,29 +84,60 @@ def plot_predicted_vs_ground_truth_density(
...
+                    slice(columns[0], columns[-1] + 1),
+                    slice(columns[0], columns[-1] + 1),
+                ]
+            )
+        fig, axs = plt.subplots(nrows=len(cms), figsize=(10, 8 * len(cms)))
+        for i, sub_cm in enumerate(cms):
+            sub_classes = classes[
+                slice(columns_set[i][0], columns_set[i][-1] + 1)
+            ]
+            axs[i].imshow(sub_cm, interpolation="nearest", cmap=plt.cm.Blues)
```
Comment (mlederbauer) [inline nmrcraft/analysis/plotting.py:130]: Super nice! in the final review, we can add our minecraft color scheme here

Comment (mlederbauer) [reply nmrcraft/analysis/plotting.py:130]: because we obtain the cmap from the setup_style() function in nmrcraft/analysis/plotting.py

File: nmrcraft/data/dataset.py [hunk 4]
```diff
@@ -225,6 +237,80 @@ def preprocess_features(self, X):
...
+        return ys_decoded_properly_rotated
+
+    def confusion_matrix_data_adapter(self, y):
+        """
+        Takes in binary encoded target array and returns decoded flat list.
+        Especially designed to work with confusion matrix.
+        """
+        y_decoded = self.binarized_target_decoder(y)
+        flat_y_decoded = [y for ys in y_decoded for y in ys]
+        return flat_y_decoded
```
Comment (mlederbauer) [inline nmrcraft/data/dataset.py:304]: That must have been technically difficult to implement! Thanks for taking care of this. In the final review, we could add a bit more detailed documentation here so that the TAs know what’s going on in this function.

File: nmrcraft/evaluation/evaluation.py [hunk 5]
```diff
@@ -1,26 +1,34 @@
+import os
 from typing import Any, Dict, Tuple
 
 from sklearn.base import BaseEstimator
 from sklearn.metrics import (
     accuracy_score,
     auc,
-    # confusion_matrix,
```
Comment (mlederbauer) [inline nmrcraft/evaluation/evaluation.py:7]: Thanks for cleaning up the code alongside!

File: nmrcraft/evaluation/evaluation.py [hunk 6]
```diff
@@ -46,3 +60,60 @@ def model_evaluation(
         fpr,
         tpr,
     )
+
+
+def model_evaluation_nD(
+    model: BaseEstimator,
```
Comment (mlederbauer) [inline nmrcraft/evaluation/evaluation.py:66]: Very nice. Also a note for final review (I’m creating an issue out of this: add type hints for all functions and classes. The purpose of this is to enforce that a function takes inputs only of a certain type, and also only outputs a certain type.

instead of 

```
def confusion_matric_plotter_or_whatever(target, full=True):
   return 1+1
```

Write

```
def confusion_matric_plotter_or_whatever(target: np.array, full: bool = True) -> int:
   return 1+1
   ```
   
   This seems a bit useless at first but its a game changer when writing code that takes in many different inputs, and saves some amount of error messages!

File: nmrcraft/evaluation/evaluation.py [hunk 7]
```diff
@@ -46,3 +60,60 @@ def model_evaluation(
...
+    X_test: Any,
+    y_test: Any,
+    y_labels: Any,
+    dataloader: dataset.DataLoader,
```
Comment (mlederbauer) [inline nmrcraft/evaluation/evaluation.py:70]: Here, for example, you have such type hints

File: scripts/training/train_metal.py [hunk 8]
```diff
@@ -66,32 +60,61 @@ def main(dataset_size, target, model_name):
...
+                title=title,
+                path=get_cm_path(),
+            )
+
+        # Logging common data
+        mlflow.log_metrics(metrics)
         mlflow.sklearn.log_model(best_model, "model")
-
         print(f"Accuracy: {metrics['accuracy']}")
+        mlflow.log_artifact(get_cm_path())
```
Comment (mlederbauer) [inline scripts/training/train_metal.py:117]: Thanks for cleaning up! We might need to adapt this to the script @strsamue  is writing.

**Review by mlederbauer — DISMISSED** [review]

**Review by mlederbauer — APPROVED** [review]
